File: src/test/java/io/github/seggan/jyxal/compiler/ElementTest.java
package io.github.seggan.jyxal.compiler;

import io.github.seggan.jyxal.runtime.RuntimeMethods;
import io.github.seggan.jyxal.runtime.math.BigComplex;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class ElementTest {

    @Test
    public void testCaseConversion() {
        Assertions.assertEquals("test", Element.screamingSnakeToCamel("TEST"));
        Assertions.assertEquals("testTest", Element.screamingSnakeToCamel("TEST_TEST"));
        Assertions.assertEquals("theLastTest", Element.screamingSnakeToCamel("THE_LAST_TEST"));
    }

    @Test
    public void assertMethodsFound() {
        for (Element element : Element.values()) {
            Element.LinkedMethodType type = element.type;
            if (type != null) {
                String methodName = Element.screamingSnakeToCamel(element.name());
                Assertions.assertDoesNotThrow(() -> {
                    //noinspection ResultOfMethodCallIgnored
                    RuntimeMethods.class.getMethod(
                            Element.screamingSnakeToCamel(element.name()),
                            type.argType
                    );
                }, "Method not found: " + methodName);
            }
        }

        RuntimeMethods.isPrime(BigComplex.valueOf(2809));
    }
}

File: src/test/kotlin/io/github/seggan/jyxal/compiler/ElementTest.kt
package io.github.seggan.jyxal.compiler

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class ElementTest {

    @Test
    fun testCaseConversion() {
        Assertions.assertEquals("test", screamingSnakeToCamel("TEST"))
        Assertions.assertEquals("testTest", screamingSnakeToCamel("TEST_TEST"))
        Assertions.assertEquals("theLastTest", screamingSnakeToCamel("THE_LAST_TEST"))
    }

    @Test
    fun assertMethodsFound() {
        for (element in Element.values()) {
            val type = element.type
            if (type != null) {
                val methodName = screamingSnakeToCamel(element.name)
                Assertions.assertDoesNotThrow<Any>({
                    Class.forName("io.github.seggan.jyxal.runtime.RuntimeMethods").getMethod(
                        screamingSnakeToCamel(element.name),
                        type.argType.java
                    )
                }, "Method not found: $methodName")
            }
        }
    }

    @Test
    fun assertNoDuplicateElements() {
        val names = Element.values().map { it.text }
        val duplicates = names.groupBy { it }.filter { it.value.size > 1 }
        Assertions.assertTrue(duplicates.isEmpty(), "Duplicate elements found: ${duplicates.keys}")
    }
}